man/kmrfsplit.1
'\"
.\" Copyright (C) 2012-2018 RIKEN R-CCS
.\" nroff -u0 -Tlp -man
.TH kmrfsplit.py 1 "2014-02-04" "KMR" "User Commands"
.SH NAME
kmrfsplit.py \- split a file
.SH SYNOPSIS
.LP
.nf
\fBkmrfsplit.py\fR [-n num] [-s pattern] [-d outdir] [-p prefix] [-f] file
.fi
.SH DESCRIPTION
.sp
.LP
kmrfsplit.py splits a file into a number of files (-n) with roughly
equal sizes.  It assumes an input file consists of many records
started by a string matching to a given pattern.  When a pattern is
not found in a certain range in the input file (around the split
point), an error will be reported.
.sp
When outdir (-d) is specified, output files are saved under the
directory.  By default, output files have names "partX" and saved in
the current directory, where X is a six-digit sequence number starting from 0
with some leading zeros.
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.na
-n num
.ad
.sp .6
.RS 4n
Specifies the number to split.  Default is 1.
.RE
.sp
.ne 2
.na
-s pattern
.ad
.sp .6
.RS 4n
Specifies a pattern string of a record separator.  A line matching to
this pattern starts a new record.  It accepts regular expressions.
Default is "\\n".
.RE
.sp
.ne 2
.na
-d outdir
.ad
.sp .6
.RS 4n
Specifies an output directory.  Output files are saved under it.
Default is the currrent directory.
.RE
.sp
.ne 2
.na
-p prefix
.ad
.sp .6
.RS 4n
Specifies a prefix to output file names.  File names are suffixed by a
number.  Default is "part".
.RE
.sp
.ne 2
.na
-f
.ad
.sp .6
.RS 4n
Forces to create an output directory if it does not exist.
.RE


CHANGELOG.md
# CHANGELOG

The list is sketchy.

## 1.9 (2018-08-27)

- 1.9 is a minor update
- Move GitHub repository from pf-aics-riken to riken-rccs
- Change the copyright holder name: RIKEN AICS to RIKEN R-CCS

## 1.8.1 (2016-04-25)

- Add ex/kmreprun.py, an example of simple shell command invoker
- Improve support of Mac OS X
- Several small bug fixes

## 1.8.0 (2015-12-25)

- Add a function tracing option for KMRViz, a KMR visualizer
- Add a simple scan function, kmr_scan_on_value()
- Add a load-leveling shuffle, kmr_shuffle_leveling_pair_count()
- Add a new kmr example, pagerank.c
- BUGFIX: kmr_match(), kmr_histogram_count_by_ranks(), kmr_find_key()

## 1.7.1 (2015-06-29)

- BUGFIX: kmr_take_one()
- BUGFIX: kmrrungenscript.py

## 1.7 (2015-06-22)

- Change license to BSD
- Strictly check options passed to kmr functions
- Add support for FOCUS supercomputer in kmrrungenscript.py
- Add Python API

## 1.6 (2015-04-01)

- Disable keep_open option in kmr functions except mapper functions
- Support no-fsync mode in Checkpoint/Restart
- Improve performance of KMRRUN when many input files are given
- BUGFIX: Checkpoint/Restart
- EXPERIMENTAL: Support selective mode that reduces IO overhead in
  Checkpoint/Restart

## 1.5 (2015-02-04)

- Reduce overhead of KMRRUN
- Temporal files for reducers are generated in subdirectories when
  using KMRRUN
- WORKAROUND: Abort on calling kmr_read_file_xxx with some MPI
  implementations
- WORKAROUND: Abort on calling kmr_map_processes and using KMRRUN
  with some Open MPI versions (1.6.3 - 1.8.1)

## 1.4 (2014-12-25)

- Resolved a scalability issue of KMRRUN and map functions that
  spawns child programs

## 1.3.2 (2014-11-07)

- Improve job script generator for KMRRUN

## 1.3.1 (2014-11-04)

- Add a word counting sample of KMRRUN
- BUGFIX: KMRRUN command line option parsing

## 1.3 (2014-09-17)

- Add a new function for affinity-aware file assignment

## 1.2 (2014-06-27)

- Support Checkpoint/Restart
- Add KMRRUN command that executes simple MapReduce workflow

## 1.1 (2013-09-20)

- 1.1 is a minor update

## 1.0 (2013-04-26)

- First release


cmd/kmrfsplit.py
#!/usr/bin/python
# Copyright (C) 2012-2018 RIKEN R-CCS

## \file kmrfsplit.py KMR-Shell File Splitter.

import sys
import os
import re
from optparse import OptionParser

##  Write part of the inputfile to outputfile.
#   @param  ipath  inputfile path.
#   @param  opath  outputfile path.
#   @startpos  start position of part.
#   @endpos    end position of part.

def writefile(ipath, opath, startpos, endpos) :
    bufsize = 0x8000000
    # read buffer size is 128Mbyte
    try:
        fin = open(ipath, "r")
    except IOError:
       print 'Error: could not open "%s".' % ipath
       sys.exit()

    try:
        fout = open(opath, "w")
    except IOError:
       print 'Error: could not open "%s".' % opath
       sys.exit()

    fin.seek(startpos, 0)
    remain = endpos - startpos
    while remain > 0 :
        # bufferd read/write.
        if bufsize > remain :
            bufsize = remain
        buf = fin.read(bufsize)
        fout.write(buf)
        remain -= len(buf)
    fin.close()
    fout.close()

##  Caluculate cutting point of file.
#   Search separator string in proper position of input file
#   and return cutting point of file.
#   If separator string not found, print error message and exit.
#
#   @param ipath     input file path.
#   @param sep       separator string. (regular expression)
#   @param startpos  start position of separate.
#   @param endpos    end position of separate.

def getcuttingpoint(ipath, sep, startpos, partsize) :
    bufsize = 0x8000000
    # read buffer size is 128Mbyte
    filesize = os.path.getsize(ipath)
    if startpos + partsize > filesize :
        # remain size of file is smaller than partition size.
        endpos = filesize
    else:
        endpos = startpos + partsize
        if endpos + bufsize > filesize :
            bufsize = filesize - endpos
        try:
            f = open(ipath, "r")
        except IOError:
            print 'Error: could not open "%s".' % ipath
            sys.exit()
        f.seek(endpos, 0)
        # read size of buffer.
        buf = f.read(bufsize)
        f.close()
        # search separator string in the buffer.
        p = re.compile(sep)
        ret = p.search(buf)
        if ret is None:
            print "Separator not found in proper position.\n"
            sys.exit()
        endpos += ret.end()
    return endpos

##  Split a file using separator string.
#
#   @param nums    number of part to split.
#   @param sep     separator string. (regular expression)
#   @param odir    output directory of splitted files.
#   @param opref   output file prefix of splitted files.
#   @param infile  input file path.

def splitfile(nums, sep, odir, opref, infile) :
    startpos = 0
    filesize = os.path.getsize(infile)
    partsize = filesize / nums

    print "Splitting file: ",
    for i in range(nums-1) :
        endpos = getcuttingpoint(infile, sep, startpos, partsize)

        # compose output file name.
        # ex: partXXXXXX, where XXXXXX is number of part.
        suffix = "%06d" % i
        opath = os.path.join(odir, (opref + suffix))
        # output cutted part of input file.
        writefile(infile, opath, startpos, endpos)
        startpos = endpos
        sys.stdout.write('.')
	sys.stdout.flush()

    # output remain part of input file.
    suffix = "%06d" % (nums-1)
    opath = os.path.join(odir, (opref + suffix))
    writefile(infile, opath, startpos, filesize)
    print "done."

## kmrfsplit main routine.
#  It works on Python 2.4 or later.

if __name__ == "__main__":

    usage = "usage: %prog [options] inputfile"
    parser = OptionParser(usage)

    parser.add_option("-n",
                      "--num-separate",
                      dest="nums",
                      type="int",
                      help="number of file separation",
                      metavar="number",
                      default=1)

    parser.add_option("-s",
                      "--separator",
                      dest="sep",
                      type="string",
                      help="separator string",
                      metavar="'string'",
                      default='\n')

    parser.add_option("-d",
                      "--output-directory",
                      dest="odir",
                      type="string",
                      help="output directory",
                      metavar="'string'",
                      default="./")

    parser.add_option("-p",
                      "--output-file-prefix",
                      dest="opref",
                      type="string",
                      help="output filename prefix",
                      metavar="'string'",
                      default="part")

    parser.add_option("-f",
                      "--force",
                      dest="force",
                      action="store_true",
                      help="force option",
                      default=False)

    (options, args) = parser.parse_args()

    # parameter check.
    if len(args) <> 1 :
        parser.error("missing parameter")
        sys.exit()

    inputfile = args[0]

    if not os.path.exists(inputfile) :
        print 'Error: inputfile %s is not exist.' % inputfile
        sys.exit()

    if os.path.exists(options.odir) :
        if not os.path.isdir(options.odir) :
            print 'Error: "%s" is not directory.' % options.odir
            sys.exit()
    else:
        if options.force :
            try:
                os.mkdir(options.odir)
            except IOError:
                print 'Error: could not create "%s".' % options.odir
                sys.exit()
        else:
            print 'Error: directory "%s" is not exist. create it or use -f option.' % options.odir
            sys.exit()

    splitfile(options.nums, options.sep, options.odir, options.opref, inputfile)

# Copyright (C) 2012-2018 RIKEN R-CCS
# This library is distributed WITHOUT ANY WARRANTY.  This library can be
# redistributed and/or modified under the terms of the BSD 2-Clause License.
